components/governance/org.wso2.carbon.governance.lcm/src/main/java/org/wso2/carbon/governance/lcm/tasks/data/JDBCLifecycleNotificationDAOImpl.java

```diff
@@ -0,0 +1,301 @@
+        queryBuilder.append(equals);
+        queryBuilder.append(questionMark);
+        return queryBuilder.toString();
+    }
+
+    /**
+     * This method to used add schedulers.
+     *
+     * @param registry              core registry.
+     * @param schedulerBean         checkpoint notification scheduler bean.
+     * @return                      true is scheduler added successfully.
+     * @throws GovernanceException  Throws when:
+     *                              <ul>
+     *                                  <li>If reading data from registry while invoking beginTransaction occurs an
+     *                                  error.</li>
+     *                                  <li>If committing transaction of adding checkpoint notification occurs an
+     *                                  error.</li>
+     *                              </ul>
+     */
+    @Override
+    public boolean addScheduler(Registry registry, LCNotification schedulerBean)
+            throws GovernanceException {
+
+        String sql = getAddSchedulerQuery();
+        try {
+            registry.beginTransaction();
+        } catch (RegistryException e) {
+            throw new GovernanceException("Error while reading data from registry while invoking beginTransaction for"
+                    + " query: " + sql, e);
+        }
+
+        JDBCDatabaseTransaction.ManagedRegistryConnection connection =
+                JDBCDatabaseTransaction.getConnection();
+        PreparedStatement preparedStatement = null;
+        try {
+            preparedStatement = connection.prepareStatement(sql);
+            preparedStatement.setString(1, schedulerBean.getRegPath());
+            preparedStatement.setString(2, schedulerBean.getLcName());
+            preparedStatement.setString(3, schedulerBean.getLcCheckpointId());
+            preparedStatement.setString(4, schedulerBean.getUUID());
+            preparedStatement.setInt(5, schedulerBean.getTenantId());
+            preparedStatement.setString(6, schedulerBean.getNotificationDate());
+            boolean result = preparedStatement.execute();
+            try {
+                registry.commitTransaction();
+                return result;
+            } catch (RegistryException e) {
+                throw new GovernanceException("Error while committing transaction of adding checkpoint notification "
+                        + "scheduler query: " + sql, e);
+            }
+        } catch (SQLException sqlException) {
+            try {
+                registry.rollbackTransaction();
+            } catch (RegistryException registryException) {
+                throw new GovernanceException("Error while transaction rollback for query: " + sql, registryException);
+            }
+            throw new GovernanceException("SQL error while creating scheduler data entry using query: " + sql,
+                    sqlException);
+        } finally {
+            if (preparedStatement != null) {
```

Comment:
Thanks for the comment. Close connection is handled in registry.commitTransaction() and registry.rollbackTransaction() methods using endTransaction() inside a finally block[1][2]. So we cannot close the connection here.

Carbon4-kernel code implementation[3] of a similar implementation is listed as a refefance.

[1] https://github.com/wso2/carbon4-kernel/blob/master/core/org.wso2.carbon.registry.core/src/main/java/org/wso2/carbon/registry/core/jdbc/dataaccess/JDBCTransactionManager.java#L128
[2] https://github.com/wso2/carbon4-kernel/blob/master/core/org.wso2.carbon.registry.core/src/main/java/org/wso2/carbon/registry/core/jdbc/dataaccess/JDBCTransactionManager.java#L178
[3]https://github.com/wso2/carbon4-kernel/blob/master/core/org.wso2.carbon.registry.core/src/main/java/org/wso2/carbon/registry/core/jdbc/dao/JDBCAssociationDAO.java#L39

Thanks